cmd/cmrel/cmd/validate_gomod.go

```diff
@@ -0,0 +1,353 @@
+	// an incorrect version of cert-manager
+	dummyCoreImportVersion = "v0.0.0-00010101000000-000000000000"
+
+	// hardLocalReplace is the path from submodules to the root of the repo. This won't
+	// necessarily always be hardcoded, but for now hardcoding it works
+	hardLocalReplace = "../../"
+)
+
+var (
+	validateGoModExample = fmt.Sprintf(`To validate a local checkout:
+
+%s %s --path <path-to-checkout>`, rootCommand, validateGoModCommand)
+)
+
+type validateGoModOptions struct {
+	// Path is the path of the cert-manager checkout to validate
+	Path string
+}
+
+func (o *validateGoModOptions) AddFlags(fs *flag.FlagSet, markRequired func(string)) {
+	fs.StringVar(&o.Path, "path", "", "Path of cert-manager checkout to validate")
+
+	markRequired("path")
+}
+
+func (o *validateGoModOptions) print() {
+	log.Printf("%s options:", validateGoModCommand)
+	log.Printf("  Path: %q", o.Path)
+}
+
+func validateGoModCmd(rootOpts *rootOptions) *cobra.Command {
+	o := &validateGoModOptions{}
+
+	cmd := &cobra.Command{
+		Use:          validateGoModCommand,
+		Short:        validateGoModDescription,
+		Long:         validateGoModLongDescription,
+		Example:      validateGoModExample,
+		SilenceUsage: true,
+		PreRun: func(cmd *cobra.Command, args []string) {
+			o.print()
+			log.Printf("---")
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runValidateGoMod(rootOpts, o)
+		},
+	}
+
+	o.AddFlags(cmd.Flags(), mustMarkRequired(cmd.MarkFlagRequired))
+
+	return cmd
+}
+
+func runValidateGoMod(rootOpts *rootOptions, o *validateGoModOptions) error {
+	var validationErrors []error
+
+	allInternalModules, err := findInternalModules(o.Path)
+	if err != nil {
+		return fmt.Errorf("failed to find all submodules in %q: %s", o.Path, err.Error())
+	}
+
+	if errs := allInternalModules.checkReplaces(); len(errs) > 0 {
+		validationErrors = append(validationErrors, errs...)
+	}
+
+	if err := allInternalModules.checkCoreModuleReplacements(); err != nil {
+		validationErrors = append(validationErrors, err)
+	}
+
+	if errs := allInternalModules.checkInternalModuleRequirements(); len(errs) > 0 {
+		validationErrors = append(validationErrors, errs...)
+	}
+
+	if errs := allInternalModules.checkGoVersions(); len(errs) > 0 {
+		validationErrors = append(validationErrors, errs...)
+	}
+
+	if len(validationErrors) > 0 {
+		log.Println("validation failed! errors:")
+		for _, err := range validationErrors {
+			log.Printf("  %s", err.Error())
+		}
+
+		return fmt.Errorf("validation failed")
+	}
+
+	return nil
+}
+
+type internalModuleList struct {
+	modules []*internalModule
+
+	coreModule *internalModule
+	submodules []*internalModule
+
+	replaceMap map[string]module.Version
+
+	// internalModuleNames is used for quickly answering the question "is this module name for an internal module". A map is used for O(1) lookup; the boolean is ignored
+	internalModuleNames map[string]bool
```

Comment:
```suggestion
	internalModuleNames map[string]struct{}
```
Can we use a `struct{}` instead of a bool here? Otherwise, the meaning of the boolean can be confusing.